packages/dds/sequence/src/revertibles.ts

```diff
@@ -134,6 +159,153 @@ export function appendLocalPropertyChangedToRevertibles(
 	return revertibles;
 }
 
+function addIfIntervalEndpoint(
+	ref: LocalReferencePosition,
+	segmentIndex: number,
+	startIntervals: { segmentIndex: number; interval: SequenceInterval }[],
+	endIntervals: { segmentIndex: number; interval: SequenceInterval }[],
+) {
+	if (refTypeIncludesFlag(ref.refType, ReferenceType.RangeBegin)) {
+		const interval = ref.properties?.interval;
+		if (interval && interval instanceof SequenceInterval) {
+			startIntervals.push({ segmentIndex, interval });
+			return true;
+		}
+	} else if (refTypeIncludesFlag(ref.refType, ReferenceType.RangeEnd)) {
+		const interval = ref.properties?.interval;
+		if (interval && interval instanceof SequenceInterval) {
+			endIntervals.push({ segmentIndex, interval });
+			return true;
+		}
+	}
+	return false;
+}
+
+function addLocalRefPlaceholders(string: SharedString, ref: LocalReferencePosition) {
+	return {
+		stay: string.createLocalReferencePosition(
+			ref.getSegment() as SharedStringSegment,
+			ref.getOffset(),
+			ReferenceType.StayOnRemove,
+			undefined,
+		),
+		slide: string.createLocalReferencePosition(
+			ref.getSegment() as SharedStringSegment,
+			ref.getOffset(),
+			ReferenceType.SlideOnRemove,
+			undefined,
+		),
+	};
+}
+
+export function appendToMergeTreeDeltaRevertibles(
+	string: SharedString,
+	deltaArgs: IMergeTreeDeltaCallbackArgs,
+	revertibles: SharedStringRevertible[],
+) {
+	if (deltaArgs.deltaSegments.length === 0) {
+		return;
+	}
+	if (deltaArgs.operation === MergeTreeDeltaType.REMOVE) {
+		const startIntervals: { segmentIndex: number; interval: SequenceInterval }[] = [];
+		const endIntervals: { segmentIndex: number; interval: SequenceInterval }[] = [];
+		let position: number | undefined;
+
+		// find interval endpoints in each segment
+		for (let i = 0; i < deltaArgs.deltaSegments.length; i++) {
+			const refs = deltaArgs.deltaSegments[i].segment.localRefs;
+			if (refs !== undefined) {
+				for (const ref of refs) {
+					addIfIntervalEndpoint(ref, i, startIntervals, endIntervals);
+				}
+
+				// get current position of deletion to search intervals later
+				const segPosition = string.getPosition(deltaArgs.deltaSegments[i].segment);
+				if (position === undefined) {
+					position = segPosition;
+				} else {
+					assert(
+						position === segPosition,
```

Comment:
I'm not sure if I understand this correctly, but would this incorrectly fire an assert for a continuous remove that happens over split segments?

Reply:
this happens after the segments were removed, and merge-tree reports such segments as having length 0. but this took me a while to get as well, so probably worth a comment lol